domains/client/domain-executor/src/domain_block_processor.rs

```diff
@@ -182,6 +196,130 @@ where
         }
     }
 
+    pub(crate) fn fork_aware_block_imports(
+        &self,
+        primary_hash: PBlock::Hash,
+        primary_number: NumberFor<PBlock>,
+    ) -> sp_blockchain::Result<ForkAwareBlockImports<Block, PBlock>> {
+        if primary_number == One::one() {
+            return Ok(ForkAwareBlockImports {
+                initial_parent: (self.client.info().genesis_hash, Zero::zero()),
+                primary_imports: vec![HashAndNumber {
+                    hash: primary_hash,
+                    number: primary_number,
+                }],
+            });
+        }
+
+        let best_hash = self.client.info().best_hash;
+        let best_number = self.client.info().best_number;
+        let best_receipt = crate::aux_schema::load_execution_receipt::<
+            _,
+            Block::Hash,
+            NumberFor<PBlock>,
+            PBlock::Hash,
+        >(&*self.client, best_hash)?
+        .ok_or_else(|| {
+            sp_blockchain::Error::Backend(format!("Execution receipt not found for {best_hash:?}"))
+        })?;
+
+        let primary_from = best_receipt.primary_hash;
+        let primary_to = primary_hash;
+
+        if primary_from == primary_to {
+            return Err(sp_blockchain::Error::Application(Box::from(
+                "Primary block {primary_hash:?} has already been processed.",
+            )));
+        }
+
+        let route =
+            sp_blockchain::tree_route(&*self.primary_chain_client, primary_from, primary_to)?;
+
+        let retracted = route.retracted();
+        let enacted = route.enacted();
+
+        tracing::trace!(
+            ?retracted,
+            ?enacted,
+            common_block = ?route.common_block(),
+            "Calculating ForkAwareBlockImports on #{best_number},{best_hash:?}"
+        );
+
+        match (retracted.is_empty(), enacted.is_empty()) {
```

Comment:
I'd use nested `if/else` here for readability. Yes, it will add one more layer of indentation, but I wouldn't need to scroll up to see what two booleans mean.

Reply:
I also had the issue where I had to keep moving up and down. One alternative is to use plain if instead, something like:

```
if retracted.is_empty() && enacted.is_empty() {
   // ...
} else if !retracted.is_empty() && enacted.is_empty() { 
   // ...
} else ...
```